### PR TITLE
add toggle to switch between ply file and original file in 3d viewer

### PR DESCRIPTION
**Pull request by ErikAmerico**
Added a toggle button to switch between viewing the ply file and original file.

### CODE REVIEW

File: src/components/AnalysisPage/ThicknessAnalysisBox/ThickAnalysisBox.jsx
Comment (ErikAmerico) [inline src/components/AnalysisPage/ThicknessAnalysisBox/ThickAnalysisBox.jsx]: if downloadPlyFilePlaceIn3dViewer fails, will setShowToggle(true) still run? or will this whole if statement just cancel?

I wasn't able to pass setShowToggle to downloadPlyFile to handle this in that function

**Comment by ErikAmerico** [comment]
Oh and we don't need to add this toggle button, I just know you mentioned having one. 

Up to you!

File: src/components/AnalysisPage/ThicknessAnalysisBox/ThickAnalysisBox.jsx
Comment (rknizzle) [reply src/components/AnalysisPage/ThicknessAnalysisBox/ThickAnalysisBox.jsx]: If downloadPlyFilePlaceIn3dViewer throws an error then setShowToggle(true) will not run. Throwing an error stops the code executing. When an error is thrown it 'bubbles up' until it goes up into a try in which case it will stop bubbling up and go into the catch block.

Heres an example of bubbling up:

```js
function top() {
  try {
    doThing()
  } catch(err) {
    console.log('catching the error, maybe there would be a UI popup or something')
  }
}

function lower() {
  lowest()
  console.log('2')
  otherThing()
}

function lowest() {
  console.log('1')
  throw new Error('a bad thing happened')
}


top()

```

This program logs '1' and then 'catching the error, maybe there would be a UI popup or something'.

It does not log '2'.

So  this shows an error bubbling up. As functions call other functions it creates new lower layers and so when one of those lower layers throws an error, it goes back up to the higher layers either until it goes into a try in which case it will stop bubbling up and go into the catch block or iif it never goes into a try block itll just bubble up until it gets to the top which would be probably the React component

**Comment by rknizzle** [comment]
@ErikAmerico I did want a toggle but now with how you made the UI easily switch back to grey when you delete a value in the threshold input I don't t hink its as necessary but it might still be good. What do you think about the toggle UI?

But I do think that I definitely don't want to release the toggle feature until we figure out how to change the color of the model without reloading the whole scene so it doesn't cause the black screen flash and the resetting of the model position. If we add the toggle I'd like it to just change the color and not reset the position of the model

**Comment by ErikAmerico** [comment]
I don't think the toggle is completely useless.

If someone has a 1 threshold value result saved and they are currently viewing a 1.7. When they wanted to switch back to the gray model, 

they wild have to delete the .7, which would briefly show the ply file for the 1 value then they can delete the 1 to get the gray model.

and I personally felt trapped when viewing the ply model. Now I feel like I'm free to look at either 😂

**Comment by ErikAmerico** [comment]
I'll start getting into switching the color without resetting the scene! And keeping the same position.

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: Can showToggle and setShowToggle be totally managed by the ModelViewer component? Does it need to be state in the Analysis page and passed around to a bunch of other components?

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:15]: Why is this a 'Ref' and why does it use the property current? Why isn't it just like modelViewerSceneHandler or something? Is there some reason why thats necessary? I'm not super familiar with what refs are in React

File: src/components/AnalysisPage/3dModelViewer/setupThreeScene.js
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/setupThreeScene.js]: Really clever solution here returning an object containing functions to modify the scene from the React component

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: Wait why are you loading and then clearing right after?

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: Just a note on commit message for cbdfbceb7b2f542f73557f64b600f202f9bbc00a:
You renamed 2 variables in the commit but only mentioned one of them in the commit message. So just be careful of that cause it can cause a little confusion. You could have either split this into 2 commits or just mentioned both renames in the commit.

I use the command `git diff --staged` very often (I created a terminal shortcut `gds` for it). It shows all the changes that you have added and are going to get commited when you run `git commit`. This can be really useful for checking whats about to be committed.

Also commits contain sort of 2 parts; a title and a message. The title is just a one line thing and then the message can be as long as you want it to be. So you could have done like:

```
Refactor: Rename variables

previousOriginalFileNameFor3dModel -> previousOriginalFileFor3dModel
&
isNewOriginalFileNameFor3dModel -> isNewOriginalFileFor3dModel
```

Where the top line is the title and the rest is the message. 

The way that you commit is just kind of only a title that extends to the next line if it gets too long.

Heres an example of a commit that has a title and a message: https://github.com/rknizzle/propermesh/commit/77ae192b77760a33c3d7f64f1fd9563f4ba844d8

The title vs message thing isn't a big deal though so you don't need to worry about changing the way you commit right now. I just wanted to introduce you to the concept

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: THis is starting to get a little confusing with all the current vs original vs previous stuff for everything. Im not quite sure how to make it better though. Maybe just refactoring to make some small well named functions or something.

Or maybe theres a way to split it apart so that instead of all the loading logic being here we could have like a function to load in a new original file and then a separate function to load the ply visualization file into the viewer.

Not sure just thinking out loud, we can keep an eye on this as we continue developing the viewer

File: src/components/AnalysisPage/3dModelViewer/setupThreeScene.js
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/setupThreeScene.js]: The toggle works a lot better now but I think theres one more optimization we can make when switching between the grey model and the thickness visualization model.  Once we load in a ply file, instead of calling the loader every time to load back in the original file to get the grey colored view, we can just change the color of the model using threejs materials. So instead of loading in the original part file every time to get the grey view we can just change the material on the mesh to show grey.

This will solve the little pause between switching out the models too. Currently if you load in a really big file it can take 4-5 seconds to switch between the views when you press the toggle because it needs to load the whole part into the loader each time.

You've already mentioned wanting to fix the little flash of the part when switching the toggle so I think this might be a good way to go about it

**Review by rknizzle — COMMENTED** [review]
Nice solution! I left a couple little changes and some questions here and then I'll merge this in when you address those. I also gave my idea for an optimization to fix the pause between showing the model when you toggle the switch. This works well how it is though so I'm also down to merge this in before you fully fix the pause between models. As soon as you answer my couple little questions and possibly make some updates for them I'm good to merge this in

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (ErikAmerico) [reply src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: yes it can, and super easily. Thanks dude!

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (ErikAmerico) [reply src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx:15]: I'm also not an expert in react Refs.

But here its a reference to the setUpThreeScene function call.

To the best of my knowledge

Using ref lets us load and clear new models into the scene without having to re-setup the scene every time a new model loads or clears. Where if we were using a modelViewerSceneHandler state, it would trigger a re-render on each change and we would be calling setUpThreeScene over and over again (like we were before i believe, causing that black flash).

.current is where the reference is located in sceneRef. Here is the output of `console.log(sceneRef)` 

image

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (ErikAmerico) [reply src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: hmmm really only because it just works this way. If i remove `sceneRef.current.clearModel()`, then the next time i load a model. the new model will be overlapping the previous model. 

But I think whats actually happening is that when this code runs, `sceneRef.current.clearModel()` doesn't actually have anything to clear because the model hasn't been added to sceneRef.current yet. Then the next time the useEffect runs, the new model will be in the process of loading while `sceneRef.current.clearModel()` gets hit and clears the previously loaded model.

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: I think just using an else here would be simpler and clearer

File: src/components/AnalysisPage/3dModelViewer/setupThreeScene.js
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/setupThreeScene.js]: Try to avoid specifying particular line numbers in code because if you add a new feature in your next PR that pushes the code down, the correct code won't line up with this comment anymore. Try to just use descriptions instead

File: src/components/AnalysisPage/3dModelViewer/setupThreeScene.js
Comment (rknizzle) [inline src/components/AnalysisPage/3dModelViewer/setupThreeScene.js]: I think this function could use a bit more detailed of a name to include what its toggling between. Just so in the future if we have like 10 more functions to modify the 3d model scene this will be more clear what its doing

**Review by rknizzle — CHANGES_REQUESTED** [review]
The switch toggle between views is working really really good now! And yea you already mentioned the couple things you need to finish up now in your commit messages.
- removing state no longer used
- getting rid of the pause/flash when you edit the threshold value

Nice detailed commit messages now too!

File: src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx
Comment (ErikAmerico) [reply src/components/AnalysisPage/3dModelViewer/3dModelViewer.jsx]: In one of my latest commits, I was able to completely remove all of this! 

no more confusion

File: src/components/AnalysisPage/3dModelViewer/setupThreeScene.js
Comment (ErikAmerico) [reply src/components/AnalysisPage/3dModelViewer/setupThreeScene.js]: This should be good now too!

**Comment by ErikAmerico** [comment]
Also when adjusting the threshold value, The gray mesh gets displayed instead of reloading the original file. 

There is still the issue of when you type in a value that already has results, It loads in the ply file and you get that black flash. 

I know we talked about it probably not getting fixed. But I feel like after working tonight, I will be able to figure it out!

**Comment by rknizzle** [comment]
Nice! Looks good merging this in

**Pull request merged**
